File: Turtle.cpp
#include "include/Turtle.h"
#include "include/Antelope.h"

Turtle::Turtle(World* world, Point& position)
    : Animal(world, 2, 1, position, "🐢") {
    // Constructor implementation
}

Turtle::Turtle(World* world, Point& position, int age)
    : Animal(world, 2, 1, position, "🐢", age) {
    // Constructor implementation
}

Turtle::~Turtle() {
    // Destructor implementation
}

bool Turtle::haveSavedAttack(Organism& other) {
    if(dynamic_cast<const Antelope*>(&other) != nullptr) {
        other.collision(*this);
        return false;
    }
    return false;
}

void Turtle::action(){
    Point destination = world->getRandomNeighbor(position);
    if(world->isWithinBounds(destination) && age != 0) {
        Organism* other = world->getAtCoordinates(destination);

        if(haveSavedAttack(*other)) {
            return;
        }
        other = world->getAtCoordinates(destination);
        int randomValue = rand() % 4 + 1;
        int status = -1;
        if(other != nullptr) {
            status = collision(*other, randomValue);
        }
        if (status != 1 && randomValue == 1) move(destination);
    }

    age++;
}

int Turtle::collision(Organism& other, int randomValue) {
    if(canKill(other) && randomValue == 1) {
        kill(other);
        return 0;
    }
    if (typeid(other) == typeid(*this)) {
        if(canReproduce(other, position)) {
            reproduce(position);
        }
        return 1;
    }
    other.collision(*this);
    return 0;
}

File: game-of-life-java/src/Turtle.java
import javax.swing.*;
import java.awt.*;
import java.util.Random;

public class Turtle extends Animal {
    public static final ImageIcon turtleIcon = new ImageIcon("images/turtle.png");
    public static final Image scaledTurtleIcon = turtleIcon.getImage().getScaledInstance(30, 30, Image.SCALE_SMOOTH);

    public Turtle(Point position, World world, int age) {
        super(2, 1, position, world, age); // Strength = 2, Initiative = 1
    }

    @Override
    public ImageIcon getIcon() {
        return turtleIcon; // Zwracanie zcache'owanej ikony
    }

    @Override
    protected boolean haveSavedAttack(Organism other) {
        if(other instanceof Antelope) {
            other.escapeCollision(this);
            return false;
        }
        return false;
    }

    @Override
    public void move() {
        Random random = new Random();
        if (random.nextInt(100) < 25) { // 25% szans na ruch
            Point newPos = world.generateRandomPosition(position);
            if (newPos != null) {
                Organism other = world.getOrganismAtPosition(newPos);
                if(haveSavedAttack(other)) {
                    return;
                }
                other = world.getOrganismAtPosition(newPos);
                if (other != null) {
                    collision(other);
                } else {
                    position = newPos; // Przesuwamy organizm na nową pozycję
                }
            }
        }
    }
}
